### PR TITLE
Integrate PagedAttention Optimization custom kernel into vLLM

### PR DESCRIPTION
**Pull request by lcskrishna**
This PR introduces the custom optimized PagedAttention on ROCm. To use the Custom PagedAttention use the following env variable ```VLLM_USE_ROCM_CUSTOM_PAGED_ATTN=1```. 

Currently this PagedAttention kernel is supported with fp16 datatype, with GQA ratio=1 to 16 and Head-size=128 and block-size=16. 

Kernel was authored by : @sanyalington 

cc: @shajrawi @sunway513

### CODE REVIEW

File: tests/kernels/test_attention_custom.py
Comment (sanyalington) [inline tests/kernels/test_attention_custom.py:1]: We should remove all the code which is not relevant to custom paged attention from this file.

**Review by sanyalington — APPROVED** [review]
Paged attention custom reduction kernel supports a max context length of 16K beyond which it can produce incorrect results. A fix for this and further reduction kernel perf improvements are going to be available in a later drop.

File: tests/kernels/test_attention_custom.py
Comment (lcskrishna) [reply tests/kernels/test_attention_custom.py:1]: Cleaned up the test script

**Comment by lcskrishna** [comment]
@shajrawi added into the documentation and enabled by default. The context length <= 32k support will be added into the next PR after this one.

**Review by shajrawi — APPROVED** [review]
lgtm

**Pull request merged**